flink-kubernetes-operator-autoscaler/src/main/java/org/apache/flink/kubernetes/operator/autoscaler/ScalingMetricCollector.java

```diff
@@ -313,41 +320,37 @@ protected Map<JobVertexID, Map<String, FlinkMetric>> queryFilteredMetricNames(
         var jobId = JobID.fromHexString(cr.getStatus().getJobStatus().getJobId());
         var vertices = topology.getVerticesInTopologicalOrder();
 
-        long deployedGeneration = getDeployedGeneration(cr);
-
-        var previousMetricNames = availableVertexMetricNames.get(ResourceID.fromResource(cr));
-
-        if (previousMetricNames != null) {
-            if (deployedGeneration == previousMetricNames.f0) {
-                // We have already gathered the metric names for this spec, no need to query again
-                return previousMetricNames.f1;
-            } else {
-                availableVertexMetricNames.remove(ResourceID.fromResource(cr));
-            }
-        }
-
-        try (var restClient = (RestClusterClient<String>) flinkService.getClusterClient(conf)) {
-            var names =
-                    vertices.stream()
-                            .collect(
-                                    Collectors.toMap(
-                                            v -> v,
-                                            v ->
-                                                    getFilteredVertexMetricNames(
-                                                            restClient, jobId, v, topology)));
-            availableVertexMetricNames.put(
-                    ResourceID.fromResource(cr), Tuple2.of(deployedGeneration, names));
-            return names;
-        }
-    }
-
-    public static long getDeployedGeneration(AbstractFlinkResource<?, ?> cr) {
-        return cr.getStatus()
-                .getReconciliationStatus()
-                .deserializeLastReconciledSpecWithMeta()
-                .getMeta()
-                .getMetadata()
-                .getGeneration();
+        var resourceId = ResourceID.fromResource(cr);
+        var names =
+                availableVertexMetricNames.compute(
+                        resourceId,
+                        (k, previousMetricNames) -> {
+                            if (previousMetricNames != null
+                                    && previousMetricNames
+                                            .keySet()
+                                            .equals(topology.getParallelisms().keySet())) {
+                                // We have already gathered the metric names for this topology
+                                return previousMetricNames;
```

Comment:
Could it be that metrics are added late and we are missing some in the first attempt?

Reply:
The logic in this sense did not change, not completely sure if this can happen or not but we only queried the metrics one per topology even before.